ops/main.py

```diff
@@ -42,6 +43,114 @@
+
+    # env var JUJU_CHARM_DIR
+    charm_dir: Optional[str] = None
+
+    # env var JUJU_DEBUG
+    debug: Optional[str] = None
+
+    # env var JUJU_DEPARTING_UNIT
+    departing_unit: Optional[str] = None
+
+    # env var JUJU_DISPATCH_PATH
+    dispatch_path: Optional[str] = None
+
+    # env var JUJU_MODEL_NAME
+    model_name: Optional[str] = None
+
+    # env var JUJU_MODEL_UUID
+    model_uuid: Optional[str] = None
+
+    # env var JUJU_NOTICE_ID
+    notice_id: Optional[str] = None
+
+    # env var JUJU_NOTICE_KEY
+    notice_key: Optional[str] = None
+
+    # env var JUJU_NOTICE_TYPE
+    notice_type: Optional[str] = None
+
+    # env var JUJU_PEBBLE_CHECK_NAME
+    pebble_check_name: Optional[str] = None
+
+    # env var JUJU_RELATION
+    relation_name: Optional[str] = None
+
+    # env var JUJU_RELATION_ID
+    relation_id: Optional[int] = None
+
+    # env var JUJU_REMOTE_APP
+    remote_app: Optional[str] = None
+
+    # env var JUJU_REMOTE_UNIT
+    remote_unit: Optional[str] = None
+
+    # env var JUJU_SECRET_ID
+    secret_id: Optional[str] = None
+
+    # env var JUJU_SECRET_LABEL
+    secret_label: Optional[str] = None
+
+    # env var JUJU_SECRET_REVISION
+    secret_revision: Optional[int] = None
+
+    # env var JUJU_STORAGE_ID
+    storage_id: Optional[str] = None
+
+    # env var JUJU_UNIT_NAME
+    unit_name: Optional[str] = None
+
+    # env var JUJU_VERSION
+    version: Optional[str] = None
+
+    # env var JUJU_WORKLOAD_NAME
+    workload_name: Optional[str] = None
+
+    @classmethod
+    def from_environ(cls, env: Dict[str, Any]) -> '_JujuContext':
+        return _JujuContext(
+            action_name=env.get('JUJU_ACTION_NAME'),
+            action_uuid=env.get('JUJU_ACTION_UUID'),
+            charm_dir=env.get('JUJU_CHARM_DIR'),
+            debug=env.get('JUJU_DEBUG'),
+            departing_unit=env.get('JUJU_DEPARTING_UNIT'),
+            dispatch_path=env.get('JUJU_DISPATCH_PATH'),
+            model_name=env.get('JUJU_MODEL_NAME'),
+            model_uuid=env.get('JUJU_MODEL_UUID'),
+            notice_id=env.get('JUJU_NOTICE_ID'),
+            notice_key=env.get('JUJU_NOTICE_KEY'),
+            notice_type=env.get('JUJU_NOTICE_TYPE'),
+            pebble_check_name=env.get('JUJU_PEBBLE_CHECK_NAME'),
+            relation_name=env.get('JUJU_RELATION'),
+            relation_id=int(env.get('JUJU_RELATION_ID', '').split(':')[-1])
+            if 'JUJU_RELATION_ID' in env
+            else None,
+            remote_app=env.get('JUJU_REMOTE_APP'),
+            remote_unit=env.get('JUJU_REMOTE_UNIT'),
+            secret_id=env.get('JUJU_SECRET_ID'),
+            secret_label=env.get('JUJU_SECRET_LABEL'),
+            secret_revision=int(env.get('JUJU_SECRET_REVISION', ''))
+            if 'JUJU_SECRET_REVISION' in env
+            else None,
+            storage_id=env.get('JUJU_STORAGE_ID'),
+            unit_name=env.get('JUJU_UNIT_NAME'),
+            version=env.get('JUJU_VERSION', '0.0.0'),
```

Comment:
We should probably not have `0.0.0` as the default here, but use None correctly here. Can do that in `JujuVersion` or where it's used if we need to.